### PR TITLE
fix: add test for and slight rewrite to hazards of imperative

### LINKED ISSUE #34394: Challenge (Hazards of Imperative Code) passes after running tests without any changes made to the code
**Describe your problem and how to reproduce it:**
After "Run the tests" is clicked, challenge is marked as completed and you can proceed to the next challenge. This happens even without making changes to the original code, meaning it happens with incorrect output.

**Add a Link to the page with the problem:**
https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/functional-programming/understand-the-hazards-of-using-imperative-code/

**Tell us about your browser and operating system:**
* Browser Name: Mozilla Firefox
* Browser Version: 63.0.3
* Operating System:  Windows 10 64-bit

**If possible, add a screenshot here (you can drag and drop, png, jpg, gif, etc. in this box):**

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #34394

This adds a simple test - checking that the final answer is the expected one.  In addition, I fixed a minor markdown issue and changed the instructions as now the user is expected to change the code.

### CODE REVIEW

**Review by thecodingaviator — DISMISSED** [review]
The PR before this one, #35052 already moves instructions to the correct place

**Comment by ojeytonwilliams** [comment]
Should I wait for that PR to get merged and then update mine accordingly?

**Comment by thecodingaviator** [comment]
@ojeytonwilliams That's what I recommend

**Comment by ojeytonwilliams** [comment]
@ojeytonwilliams Will do. Thanks!

**Comment by RandellDawson** [comment]
@ojeytonwilliams PR #35052 has been merged.  If you think your changes are still relevant, then fix the merge conflict and we can take a look at your changes.

Thanks.

**Comment by ojeytonwilliams** [comment]
@thecodingaviator @RandellDawson I've updated my PR to fix the conflict and slightly tweak the wording.  Apologies for the extra commit!

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/understand-the-hazards-of-using-imperative-code.english.md [hunk 1]
```diff
@@ -14,21 +14,25 @@ JavaScript offers many predefined methods that handle common tasks so you don't
...
 </section>
 
 ## Tests
 <section id='tests'>
 
 ```yml
 tests:
-  - text: Move ahead to understand the error.
-    testString: '' # assert(true, 'Move ahead to understand the error.');
+  - text: finalTabs.tabs should be ['FB', 'Gitter', 'Reddit', 'Twitter', 'Medium', 'new tab', 'Netflix', 'YouTube', 'Vine', 'GMail', 'Work mail', 'Docs', 'freeCodeCamp', 'new tab']
```
Comment (thecodingaviator) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/understand-the-hazards-of-using-imperative-code.english.md]:
```suggestion
  - text: <code>finalTabs.tabs</code> should be <code>['FB', 'Gitter', 'Reddit', 'Twitter', 'Medium', 'new tab', 'Netflix', 'YouTube', 'Vine', 'GMail', 'Work mail', 'Docs', 'freeCodeCamp', 'new tab']</code>
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/understand-the-hazards-of-using-imperative-code.english.md [hunk 2]
```diff
@@ -90,6 +99,41 @@ alert(finalTabs.tabs);
 <section id='solution'>
 
 ```js
```
Comment (thecodingaviator) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/understand-the-hazards-of-using-imperative-code.english.md]: @RandellDawson Shouldn't this be:
````
```html
````
?

**Review by thecodingaviator — CHANGES_REQUESTED** [review]

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/understand-the-hazards-of-using-imperative-code.english.md (hunk 2 above)
Comment (ojeytonwilliams) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/understand-the-hazards-of-using-imperative-code.english.md]: @thecodingaviator This challenge is pure js, so I think it's fine as is.

Comment (thecodingaviator) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/understand-the-hazards-of-using-imperative-code.english.md]: You're right, my bad

**Review by RandellDawson — APPROVED** [review]
LGTM  @thecodingaviator You still show you have changes requested for this PR.  If you are fine with the changes, then approve and merge the PR.

**Review by moT01 — CHANGES_REQUESTED** [review]
I did some digging here because I couldn't get this to pass. Your solution...
```
  var tabsBeforeIndex = this.tabs.splice(0, index); // get the tabs before the tab
  var tabsAfterIndex = this.tabs.splice(index + 1); // get the tabs after the tab
```
This won't work because `this.tabs` is changed after running the first .splice - so the first time this is run on line 44 (`videoWindow.tabClose(2)`) - the videoWindow.tabs array changes from `['Netflix', 'YouTube', 'Vimeo', 'Vine']`, to just `['Vimeo', 'Vine']` after the first splice. And that is what the second splice has to work with. 

Changing the two .splice methods to .slice should fix this - as .slice won't mutate the original array.

**Comment by ojeytonwilliams** [comment]
@moT01 Yeah, the seed is supposed to fail for exactly the reasons you mentioned.  The idea being that the camper comes to the same conclusion you did, rewrites it to use `slice` and passes the tests.

However, now that I've revisited this, I see that the original code wasn't just wrong because it mutated `this.tabs`.  It should be what you wrote:
```js
var tabsBeforeIndex = this.tabs.splice(0, index); // get the tabs before the tab
var tabsAfterIndex = this.tabs.splice(index + 1); // get the tabs after the tab
```
but the +1 is missing from the current version.  I've fixed that, since we want the only problem to be the mutation.

**Comment by moT01** [comment]
I was thinking you wanted the user to the `+1` since the challenge mentions - "`This helps to avoid semantic errors, like the "Off By One Errors" that were covered in the Debugging section.`" - but either way works

**Review by moT01 — APPROVED** [review]

**Comment by ojeytonwilliams** [comment]
@moT01 Indeed - the challenge description does hint in that direction.  Hopefully this won't be too confusing, though.

**Comment by moT01** [comment]
Looks good @ojeytonwilliams! 🎉

**Pull request merged**